rrtypes.go

```diff
@@ -0,0 +1,190 @@
+package libdns
+
+import (
+	"fmt"
+	"net/netip"
+	"time"
+)
+
+// Address represents a parsed A-type or AAAA-type record,
+// which associates a name with an IPv4 or IPv6 address
+// respectively. This is typically how to "point a domain
+// to your server."
+//
+// Since A and AAAA are semantically identical, with the
+// exception of the bit length of the IP address in the
+// data field, these record types are combined for ease of
+// use in Go programs, which supports both address sizes,
+// to help simplify code.
+type Address struct {
+	Name string
+	TTL  time.Duration
+	IP   netip.Addr
+}
+
+func (a Address) RR() RR {
+	recType := "A"
+	if a.IP.Is6() {
+		recType = "AAAA"
+	}
+	return RR{
+		Name: a.Name,
+		TTL:  a.TTL,
+		Type: recType,
+		Data: a.IP.String(),
+	}
+}
+
+// CAA represents a parsed CAA-type record, which is used to specify which PKIX
+// certificate authorities are allowed to issue certificates for a domain. See
+// also the [registry of flags and tags].
+//
+// [registry of flags and tags]: https://www.iana.org/assignments/caa-parameters/caa-parameters.xhtml
+type CAA struct {
+	Name  string
+	TTL   time.Duration
+	Flags uint8 // As of March 2025, the only valid values are 0 and 128.
+	Tag   string
+	Value string
+}
+
+func (c CAA) RR() RR {
+	return RR{
+		Name: c.Name,
+		TTL:  c.TTL,
+		Type: "CAA",
+		Data: fmt.Sprintf(`%d %s %q`, c.Flags, c.Tag, c.Value),
+	}
+}
+
+// CNAME represents a CNAME-type record, which delegates
+// authority to other names.
+type CNAME struct {
+	Name   string
+	TTL    time.Duration
+	Target string
+}
+
+func (c CNAME) RR() RR {
+	return RR{
+		Name: c.Name,
+		TTL:  c.TTL,
+		Type: "CNAME",
+		Data: c.Target,
+	}
+}
+
+// HTTPS represents a parsed HTTPS-type record, which is used
+// to provide clients with information for establishing HTTPS
+// connections to servers. It may include data about ALPN,
+// ECH, IP hints, and more.
+type HTTPS struct {
+	Name     string
+	TTL      time.Duration
+	Priority uint16
+	Target   string
+	Value    SvcParams
+}
+
+// RR converts the parsed record data to a generic [Record] struct.
+//
+// EXPERIMENTAL; subject to change or removal.
+func (h HTTPS) RR() RR {
+	return RR{
+		Name: h.Name,
+		TTL:  h.TTL,
+		Type: "HTTPS",
+		Data: fmt.Sprintf("%d %s %s", h.Priority, h.Target, h.Value),
+	}
+}
+
+// MX represents a parsed MX-type record, which is used to specify the hostnames
+// of the servers that accept mail for a domain.
+type MX struct {
+	Name       string
+	TTL        time.Duration
+	Preference uint16 // Lower values indicate that clients should prefer this server
```

Comment:
What should this field be called? The RFCs call it `Preference`, but since `SRV` and `HTTPS` use `Priority`, I'm thinking that maybe we should call it `Priority` here for consistency?

Reply:
Personally I'd prefer `Preference` since that's what the RFC for MX calls it.

The name doesn't have to be the same as in the other types, since this is a concrete type that requires its own code anyways, so it's not like an interface where it has to be the same for all types that implement it. :+1: